chainntnfs/bitcoindnotify/bitcoind.go

```diff
@@ -385,60 +390,123 @@ out:
-	// then we may be able to dispatch it immediately.
-	// TODO: fall back to scanning blocks if txindex isn't on.
 	tx, err := b.chainConn.GetRawTransactionVerbose(txid)
-	if err != nil || tx == nil || tx.BlockHash == "" {
-		if err == nil {
-			return nil, nil
-		}
-		// Do not return an error if the transaction was not found.
-		if jsonErr, ok := err.(*btcjson.RPCError); ok {
-			if jsonErr.Code == btcjson.ErrRPCNoTxInfo {
-				return nil, nil
-			}
+	if err != nil {
+		// Avoid returning an error if the transaction was not found.
+		jsonErr, ok := err.(*btcjson.RPCError)
+		if !ok || jsonErr.Code != btcjson.ErrRPCNoTxInfo {
+			return nil, fmt.Errorf("unable to query for txid "+
+				"%v: %v", txid, err)
 		}
-		return nil, fmt.Errorf("unable to query for txid(%v): %v", txid, err)
+	}
+
+	// Make sure we actually retrieved a transaction that is included in a
+	// block. Without this, we won't be able to retrieve its confirmation
+	// details.
+	if tx == nil || tx.BlockHash == "" {
+		return nil, nil
 	}
 
 	// As we need to fully populate the returned TxConfirmation struct,
 	// grab the block in which the transaction was confirmed so we can
 	// locate its exact index within the block.
 	blockHash, err := chainhash.NewHashFromStr(tx.BlockHash)
 	if err != nil {
-		return nil, fmt.Errorf("unable to get block hash %v for historical "+
-			"dispatch: %v", tx.BlockHash, err)
+		return nil, err
 	}
+
 	block, err := b.chainConn.GetBlockVerbose(blockHash)
 	if err != nil {
-		return nil, fmt.Errorf("unable to get block hash: %v", err)
+		return nil, fmt.Errorf("unable to get block with hash %v: %v",
+			blockHash, err)
 	}
 
-	// If the block obtained, locate the transaction's index within the
+	// If the block was obtained, locate the transaction's index within the
 	// block so we can give the subscriber full confirmation details.
-	txIndex := -1
 	targetTxidStr := txid.String()
-	for i, txHash := range block.Tx {
+	for txIndex, txHash := range block.Tx {
 		if txHash == targetTxidStr {
-			txIndex = i
-			break
+			return &chainntnfs.TxConfirmation{
+				BlockHash:   blockHash,
+				BlockHeight: uint32(block.Height),
+				TxIndex:     uint32(txIndex),
+			}, nil
 		}
 	}
 
-	if txIndex == -1 {
-		return nil, fmt.Errorf("unable to locate tx %v in block %v",
-			txid, blockHash)
-	}
+	// We return an error because we should have found the transaction
+	// within the block, but didn't.
+	return nil, fmt.Errorf("unable to locate tx %v in block %v", txid,
+		blockHash)
+}
+
+// confDetailsManually looks up whether a transaction is already included in a
+// block in the active chain by scanning the chain's blocks, starting from the
+// earliest height the transaction could have been included in, to the current
+// height in the chain. If it is found, the confirmation details are returned.
+// Otherwise, an error is not returned.
+func (b *BitcoindNotifier) confDetailsManually(txid *chainhash.Hash,
```

Comment:
With this change, we'll need to do a _clean sweep_ through the _entire_ codebase to ensure that `heightHint` is set  properly. There are some known cases with neutrino (particularly in the `contractcourt`) where an incorrect height hint value can cause us to actually miss a notification all together. We'll need to extensively test this patch to ensure these holes are patched up.

Reply:
I was running this patch on testnet for a bit and didn't run into any issues. I'll continue testing and see if I can run into any more.